core/matching/matching_strategy_runner.go

```diff
@@ -35,7 +39,7 @@ func MatchingStrategyRunner(req models.RequestDetails, webserver bool, simulatio
 
 		strategy.Matching(StateMatcher(state, requestMatcher.RequiresState), "state")
```

Comment:
Good optimization to remove the copy map from the loop. However I want to drop a note about the race condition here. The state could be altered by another thread at any one point during the matching, and the matching error message could be saying Hoverfly has the state of `foo=bar` and the required state is also `foo=bar` which is contradicting.

I think the whole matching operation should be synchronized to avoid the race condition, it's not safe to let other thread to modify the states while matching is in progress. I can't tell what the performance implication is, and we should fix it at some point.

Reply:
We can use copyState map (that we create before loop starts) as the parameter in StateMatcher.

Reply:
That's true. Would you like to submit another PR?

Reply:
Sure